parquet-column/src/main/java/org/apache/parquet/schema/Types.java

```diff
@@ -555,114 +622,1267 @@ protected GroupType build(String name) {
+      this.keyType = keyType;
+    }
+
+    protected void setValueType(Type valueType) {
+      this.valueType = valueType;
+    }
+
+    private Type keyType = null;
+    private Type valueType = null;
+    private Class<P> returnType = null;
+
+    public BaseMapBuilder(P parent) {
+      super(parent);
+    }
+
+    private BaseMapBuilder(Class<P> returnType) {
+      super(returnType);
+      this.returnType = returnType;
+    }
+
+    @Override
+    protected abstract Q self();
+
+    public MapKeyBuilder<P, Q> key(PrimitiveTypeName type) {
+      if (parent != null) {
+        return new MapKeyBuilder<P, Q>(this, type);
+      } else {
+        return new MapKeyBuilder<P, Q>(this, returnType, type);
+      }
+    }
+
+    public Q key(Type type) {
+      setKeyType(type);
+      return self();
+    }
+
+    public MapGroupKeyBuilder<P, Q> groupKey() {
+      return new MapGroupKeyBuilder<P, Q>(this);
+    }
+
+    public MapListKeyBuilder<P, Q> listKey() {
+      return new MapListKeyBuilder<P, Q>(this);
+    }
+
+    public MapValueBuilder<P, Q> value(PrimitiveTypeName type, Type.Repetition repetition) {
+      if (parent != null) {
+        return new MapValueBuilder<P, Q>(this,
+            type).repetition(repetition);
+      } else {
+        return new MapValueBuilder<P, Q>(this,
+            returnType, type).repetition(repetition);
+      }
+    }
+
+    public MapValueBuilder<P, Q> requiredValue(PrimitiveTypeName type) {
+      return value(type, Type.Repetition.REQUIRED);
+    }
+
+    public MapValueBuilder<P, Q> optionalValue(PrimitiveTypeName type) {
+      return value(type, Type.Repetition.OPTIONAL);
+    }
+
+    public MapGroupValueBuilder<P, Q> groupValue(Type.Repetition repetition) {
+      setKeyType(build("key"));
+      return new MapGroupValueBuilder<P, Q>(this).repetition(repetition);
+    }
+
+    public MapGroupValueBuilder<P, Q> requiredGroupValue() {
+      return groupValue(Type.Repetition.REQUIRED);
+    }
+
+    public MapGroupValueBuilder<P, Q> optionalGroupValue() {
+      return groupValue(Type.Repetition.OPTIONAL);
+    }
+
+    public MapMapValueBuilder<P, Q> mapValue(Type.Repetition repetition) {
+      setKeyType(build("key"));
+      return new MapMapValueBuilder<P, Q>(this).repetition(repetition);
+    }
+
+    public MapMapValueBuilder<P, Q> requiredMapValue() {
+      return mapValue(Type.Repetition.REQUIRED);
+    }
+
+    public MapMapValueBuilder<P, Q> optionalMapValue() {
+      return mapValue(Type.Repetition.OPTIONAL);
+    }
+
+    public MapListValueBuilder<P, Q> listValue(Type.Repetition repetition) {
+      setKeyType(build("key"));
+      return new MapListValueBuilder<P, Q>(this).repetition(repetition);
+    }
+
+    public MapListValueBuilder<P, Q> requiredListValue() {
+      return listValue(Type.Repetition.REQUIRED);
+    }
+
+    public MapListValueBuilder<P, Q> optionalListValue() {
+      return listValue(Type.Repetition.OPTIONAL);
+    }
+
+    public Q value (Type type) {
```

Comment:
Nit: style is off for these methods: shouldn't have a space before the open paren.